### ISSUE TITLE
Differentiation rule for tridiagonal_solve

### ISSUE THREAD

**Issue by segasai** [jax-ml/jax#25693]
Hi, 

I don't think it's bug report, but a feature request. I tried to recently switch away from my own implementation of tridiagonal_solve using Thomas algorithm to the jax.lax.tridiagonal_solve and I discovered that the tridiagonal_solve implementation in jax does not seem to support differentiation. The code (given below) gives this error:
```
  File "/home/koposov/pyenv310/lib/python3.10/site-packages/jax/_src/core.py", line 468, in bind_with_trace
    return trace.process_primitive(self, args, params)
  File "/home/koposov/pyenv310/lib/python3.10/site-packages/jax/_src/interpreters/ad.py", line 395, in process_primitive
    raise NotImplementedError(msg)
NotImplementedError: Differentiation rule for 'tridiagonal_solve' not implemented
```

I also did not see any mention of not supported differentiation in the docs. 

I use python 3.10, jax 0.38 on CPU. 

If the error if not an oversight, and it is not too difficult to implement differentiation tridiagonal_solve, I could maybe take a look at doing that, if I get some pointers where to look.

Thanks

The reproducer (together with my own implementation of tridiagonal solve that does support differentiation). 

```python
import jax
import jax.numpy as jnp
import numpy as np
import jax.lax.linalg as jll


def func1(carry, x):
    # forward loop of the thomas algo
    oldc_, oldd_ = carry
    a, b, c, d = x
    cdash = c / (b - a * oldc_)
    ddash = (d - a * oldd_) / (b - a * oldc_)
    return (cdash, ddash), (cdash, ddash)


def func2(carry, xx):
    # backwards loop of thomas algo
    xold = carry
    c, d = xx
    xnew = (d - c * xold)
    return xnew, xnew


def solver(a, b, c, d):
    """
    Solve A x = d where A consists of (a,b,c) on lower/mid/upper diagonal
    uses Thomas algorithm 
    """
    cdnew = jax.lax.scan(func1, (0, 0),
                         jnp.transpose(jnp.vstack((a, b, c, d)), (1, 0)))[1]
    xnew = jax.lax.scan(
        func2, (0),
        jnp.transpose(jnp.vstack((cdnew[0], cdnew[1])), (1, 0))[::-1, :])[-1]
    return xnew[::-1]


if __name__ == '__main__':
    np.random.seed(43)
    a, b, c, d = np.random.normal(size=(4, 3))
    a[0] = 0
    c[-1] = 0
    print('x', solver(a, b, c, d))
    a, b, c, d = [jnp.array(_) for _ in [a, b, c, d]]
    print('y', jll.tridiagonal_solve(a, b, c, d[:, None])[:, 0])

    def func(x):
        ret = jnp.sum(jll.tridiagonal_solve(a, b, c, x[:, None]))
        return ret

    JG = jax.grad(func, [0])
    JG(d)

```

**Comment by patrick-kidger** [comment]
Take a look at [Lineax](https://github.com/patrick-kidger/lineax), which should support this :)

**Comment by dfm** [comment]
@segasai — Thanks for bringing this up! I agree that it makes sense for JAX to support this directly, and I don't expect it would be too complicated to add. If you're keen to make a PR, that would be awesome and I'll add some pointers below. I'm also happy to take a stab at adding it, depending on your level of enthusiasm :D 

To add AD support to that primitive, you'll need to add a JVP rule (it looks like there is already a transpose rule!) here:

https://github.com/jax-ml/jax/blob/dbe9ccd6dccd83c365021677c7e17e843d4559c4/jax/_src/lax/linalg.py#L2457-L2464

There's a TODO there for @tomhennigan to add AD using [`jax.lax.custom_linear_solve`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.custom_linear_solve.html) (I don't expect it's high on his to-do list :D), but I'd probably just implement the JVP rule directly. Take a look at the JVP rule for `triangular_solve` here:

https://github.com/jax-ml/jax/blob/dbe9ccd6dccd83c365021677c7e17e843d4559c4/jax/_src/lax/linalg.py#L1310-L1312

and here:

https://github.com/jax-ml/jax/blob/dbe9ccd6dccd83c365021677c7e17e843d4559c4/jax/_src/lax/linalg.py#L1230-L1261

for an idea of what that might look like.

Tests would go somewhere close to here:

https://github.com/jax-ml/jax/blob/dbe9ccd6dccd83c365021677c7e17e843d4559c4/tests/linalg_test.py#L1787

Again, here's where the AD tests for triangular solve live as a reference:

https://github.com/jax-ml/jax/blob/dbe9ccd6dccd83c365021677c7e17e843d4559c4/tests/linalg_test.py#L1610-L1643

Let me know what you think!

**Comment by segasai** [comment]
Thanks for the suggestions @dfm ! 
I think it it will probably take quite a bit of time to figure it out (I'm quite busy till end of Jan), so if you can do it quicker, that'd be great, otherwise, I'll try to slowly look into this.

**Comment by dfm** [comment]
Here's my update:

I've added a JVP rule [on a branch](https://github.com/jax-ml/jax/compare/main...dfm:jax:tri-diag-jvp) and discovered a few issues with the existing implementation along the way. I need to do a few more tests then I'll open the PR tomorrow!

**Comment by segasai** [comment]
Cool! Thanks! I can see that you coded the Thomas algorithm now. Before I couldn't quite recognize it anywhere.

**Comment by dfm** [comment]
> Cool! Thanks! I can see that you coded the Thomas algorithm now. Before I couldn't quite recognize it anywhere.

Yeah, it was there before too, but obscured because of how it handled batch dimensions (and some bugs!). On GPU, this gets lowered away to cuSPARSE, but on CPU that is the code that is executed.
